### PR TITLE
MOD-5948: Respect timeout policy P1 (single shard)

### PR DESCRIPTION
**Pull request by raz-mon**
Currently, the `ON_TIMEOUT` configuration parameter isn't being respected well enough, mainly due to the following:
The main problem today, is that the `sendChunk` function populates the response to the client 'on the fly', i.e., as results from the pipeline are received. This is problematic, since there is no discarding API for a Redis response, such that if we experience a timeout throughout pipeline execution, we can no longer report this to the client (in Resp3 we can report it, but it will have to be accompanied by the results that were already serialized to the response, which is incompatible with our desires of the `ON_TIMEOUT FAIL` timeout policy).

In order to respect the `ON_TIMEOUT FAIL` timeout policy, we aggregate the results from the pipeline prior to populating the response to the client, such that we can reply with an error/the results respectively to whether we experienced a timeout throughout the execution of the whole command or not. This is also a good opportunity to check for timeout, since we cover the whole pipeline execution, rather than only the first phase of it, which is what we cover today - thus this is added as well.

* Note 1: We are aware that this is the "aggressive" approach to this fix, i.e., one that may yield serious performance regressions. The decision was made mainly due to the 2 next factors:
  1. This is a **first-phase** modification. As discussed with @MeirShpilraien, we should have the response-discarding API for the Redis response for redis 8, such that we will be able to remove all self-made aggregations (introduced by this PR) and by thus all performance regressions. Thus, this is a temporary hit only.
  2. The performance regressions are experienced ONLY when the timeout policy is `FAIL`, i.e., no performance hits will be experienced to customers that don't want the strict timeout behavior.
  3. We have the open window to relax the potential hit with numerous optimizations, such as performing the mentioned aggregation only when we have no depleting results-processors in the pipeline, etc.. Today, we choose this implementation for maximal timeout coverage (for maximal user-credibility), and unity among the different commands' behavior and timeout reporting qualities.

* Note 2: This PR handles the single-shard build only. A following one will handle the cluster build (coordinator).

Also, this PR refactors the `sendChunk` function, which is complex and inefficient (and incorrect in several places).

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

**Comment by filipecosta90** [comment]
### Automated performance analysis summary

This comment was automatically generated given there is performance data available.

In summary:
- Detected a total of 7 stable tests between versions.
- Detected a total of 1 highly unstable benchmarks.
- Detected a total of 2 improvements above the improvement water line.
- Detected a total of 1 regressions bellow the regression water line 5.0.

You can check a comparison in detail via the [grafana link](https://benchmarksrediscom.grafana.net/d//3Ejv2wZnk/?var-branch=master&var-branch=razmon-respect_timeout_policy_single_shard)

### Comparison between master and razmon-respect_timeout_policy_single_shard.

Time Period from 30 days ago. (environment used: oss-standalone)

|                              Test Case                               |Baseline master (median obs. +- std.dev)|Comparison razmon-respect_timeout_policy_single_shard (median obs. +- std.dev)|% change (higher-better)|                Note                |
|----------------------------------------------------------------------|----------------------------------------|-----------------------------------------------------------------------------:|------------------------|------------------------------------|
|ftsb-10K-enwiki_abstract-hashes-term-suffix                           | 2082 +- 0.2% (7 datapoints)            |                                                                          1975|-5.1%                   |REGRESSION                          |
|ftsb-1M-enwiki_abstract-hashes-fulltext-2word-union-query-non-sortable| 996 +- 7.5% (7 datapoints)             |                                                                          1093|9.7%                    |waterline=7.5%. IMPROVEMENT         |
|ftsb-1M-nyc_taxis-hashes-load                                         | 23058 +- 1.8% (7 datapoints)           |                                                                         23320|1.1%                    |No Change                           |
|search-filtering-tag-numeric-filter-pipeline                          | 18722 +- 3.3% (7 datapoints)           |                                                                         18492|-1.2%                   |No Change                           |
|search-ftsb-10K-enwiki_abstract-hashes-term-withsuffix-trie           | 38640 +- 3.3% (7 datapoints)           |                                                                         41423|7.2%                    |IMPROVEMENT                         |
|search-ftsb-5200K-docs-union-iterators-q1                             | 1 +- 8.4% (7 datapoints)               |                                                                             1|-3.2%                   |waterline=8.4%. potential REGRESSION|
|search-numeric-optimize                                               | 8976 +- 1.1% (7 datapoints)            |                                                                          8958|-0.2%                   |No Change                           |
|search-numeric-sortby-desc-optimize                                   | 51 +- 13.1% UNSTABLE (7 datapoints)    |                                                                            60|17.9%                   |UNSTABLE (very high variance)       |
|vecsim-arxiv-titles-384-angular-filters-m16-ef-128-fulltext-filter    | 490 +- 1.6% (7 datapoints)             |                                                                           483|-1.5%                   |No Change                           |
|vecsim-arxiv-titles-384-angular-filters-m16-ef-128-numeric-filter     | 109 +- 0.6% (7 datapoints)             |                                                                           107|-1.4%                   |No Change                           |
|vecsim-arxiv-titles-384-angular-filters-m16-ef-128-tag-filter         | 58216 +- 4.1% (7 datapoints)           |                                                                         59513|2.2%                    |No Change                           |

**Review by nafraf — APPROVED** [review]

**Review by oshadmi — COMMENTED** [review]
Some comments

File: src/aggregate/aggregate_request.c
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/aggregate/aggregate_request.c:1]: What do you mean? There is a small change here below.

File: src/aggregate/aggregate_exec.c [hunk 1]
```diff
@@ -293,208 +293,284 @@ static size_t getResultsFactor(AREQ *req) {
...
-               req->optimizer->type != Q_OPT_NO_SORTER) {
-      PLN_ArrangeStep *arng = AGPLN_GetArrangeStep(&req->ap);
-      size_t reqLimit = arng && arng->isLimited? arng->limit : DEFAULT_LIMIT;
-      size_t reqOffset = arng && arng->isLimited? arng->offset : 0;
-      size_t resultFactor = getResultsFactor(req);
-
-      size_t expected_res = reqLimit + reqOffset <= req->maxSearchResults ? req->qiter.totalResults : MIN(req->maxSearchResults, req->qiter.totalResults);
-      size_t reqResults = expected_res > reqOffset ? expected_res - reqOffset : 0;
-
-      resultsLen = MIN(limit, MIN(reqLimit, reqResults));
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/aggregate/aggregate_exec.c:339]: It's not good, but ok I'll leave it. We should still open a ticket and refactor this later on.

**Review by oshadmi — APPROVED** [review]
Reported performance degredations are not due to this PR (since `ON_TIMEOUT` is not `FAIL` - benchmark is using the default module configuration)

**Pull request merged**